File: packages/sdks/server-sdks/gatrix-go-server-sdk/types/feature_flags.go
package types

// FeatureFlag represents a feature flag definition
type FeatureFlag struct {
	ID            string            `json:"id"`
	Name          string            `json:"name"`
	IsEnabled     bool              `json:"isEnabled"`
	Strategies    []FeatureStrategy `json:"strategies,omitempty"`
	Variants      []Variant         `json:"variants,omitempty"`
	ValueType     string            `json:"valueType,omitempty"`     // string, number, boolean, json
	EnabledValue  interface{}       `json:"enabledValue,omitempty"`  // Value when enabled (no strategy match)
	DisabledValue interface{}       `json:"disabledValue,omitempty"` // Value when disabled
	ValueSource   string            `json:"valueSource,omitempty"`   // environment or flag
}

// FeatureStrategy represents a targeting strategy
type FeatureStrategy struct {
	Name        string              `json:"name"`
	Parameters  *StrategyParameters `json:"parameters,omitempty"`
	Constraints []Constraint        `json:"constraints,omitempty"`
	Segments    []string            `json:"segments,omitempty"` // Segment name references
	IsEnabled   bool                `json:"isEnabled"`
}

// StrategyParameters holds rollout parameters
type StrategyParameters struct {
	Rollout    *float64 `json:"rollout,omitempty"`
	Stickiness string   `json:"stickiness,omitempty"`
	GroupID    string   `json:"groupId,omitempty"`
}

// Constraint represents a targeting constraint
type Constraint struct {
	ContextName     string   `json:"contextName"`
	Operator        string   `json:"operator"` // ConstraintOperator
	Value           string   `json:"value,omitempty"`
	Values          []string `json:"values,omitempty"`
	CaseInsensitive bool     `json:"caseInsensitive,omitempty"`
	Inverted        bool     `json:"inverted,omitempty"`
}

// Variant represents a flag variant
type Variant struct {
	Name    string      `json:"name"`
	Weight  int         `json:"weight"`
	Value   interface{} `json:"value,omitempty"`
	Enabled bool        `json:"enabled"`
}

// FeatureSegment represents a feature segment
type FeatureSegment struct {
	ID          string       `json:"id"`
	Name        string       `json:"name"`
	IsActive    bool         `json:"isActive"`
	Constraints []Constraint `json:"constraints,omitempty"`
}

// EvaluationContext holds user/session context for flag evaluation
type EvaluationContext struct {
	UserID        string                 `json:"userId,omitempty"`
	SessionID     string                 `json:"sessionId,omitempty"`
	AppName       string                 `json:"appName,omitempty"`
	AppVersion    string                 `json:"appVersion,omitempty"`
	RemoteAddress string                 `json:"remoteAddress,omitempty"`
	Properties    map[string]interface{} `json:"properties,omitempty"`
}

// EvaluationReason is the reason for evaluation result
type EvaluationReason string

const (
	ReasonDisabled      EvaluationReason = "disabled"
	ReasonDefault       EvaluationReason = "default"
	ReasonStrategyMatch EvaluationReason = "strategy_match"
	ReasonNotFound      EvaluationReason = "not_found"
)

// EvaluationResult holds the result of a flag evaluation
type EvaluationResult struct {
	ID       string           `json:"id"`
	FlagName string           `json:"flagName"`
	Enabled  bool             `json:"enabled"`
	Reason   EvaluationReason `json:"reason"`
	Variant  *Variant         `json:"variant,omitempty"`
}

// EvaluationDetail holds evaluation result with details
type EvaluationDetail[T any] struct {
	Value       T                `json:"value"`
	Reason      EvaluationReason `json:"reason"`
	FlagName    string           `json:"flagName"`
	VariantName string           `json:"variantName,omitempty"`
}

// VALUE_SOURCE constants
const (
	ValueSourceEnvDefaultEnabled   = "env:default:enabled"
	ValueSourceEnvDefaultDisabled  = "env:default:disabled"
	ValueSourceFlagDefaultEnabled  = "flag:default:enabled"
	ValueSourceFlagDefaultDisabled = "flag:default:disabled"
	ValueSourceMissing             = "missing"
)

// FlagMetric represents a single flag evaluation metric
type FlagMetric struct {
	EnvironmentID string `json:"environmentId"`
	FlagName      string `json:"flagName"`
	Enabled       bool   `json:"enabled"`
	VariantName   string `json:"variantName,omitempty"`
}

// FlagMetricsPayload is the payload sent to the metrics endpoint
type FlagMetricsPayload struct {
	Metrics    []AggregatedMetric `json:"metrics"`
	Bucket     MetricsBucket      `json:"bucket"`
	Timestamp  string             `json:"timestamp"`
	SDKVersion string             `json:"sdkVersion"`
}

// AggregatedMetric represents an aggregated metric
type AggregatedMetric struct {
	FlagName    string `json:"flagName"`
	Enabled     bool   `json:"enabled"`
	VariantName string `json:"variantName,omitempty"`
	Count       int    `json:"count"`
}

// MetricsBucket represents a time window for metrics
type MetricsBucket struct {
	Start string `json:"start"`
	Stop  string `json:"stop"`
}

// FeatureFlagsAPIResponse is the API response for GET /api/v1/server/features
type FeatureFlagsAPIResponse struct {
	Flags     []FeatureFlag    `json:"flags"`
	Segments  []FeatureSegment `json:"segments"`
	ProjectID string           `json:"projectId,omitempty"`
}

File: packages/sdks/server-sdks/gatrix-rust-server-sdk/src/types/feature_flags.rs
// Gatrix Rust Server SDK
// Feature flag types for evaluation

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Feature flag value type
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ValueType {
    String,
    Number,
    Boolean,
    Json,
}

impl std::fmt::Display for ValueType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ValueType::String => write!(f, "string"),
            ValueType::Number => write!(f, "number"),
            ValueType::Boolean => write!(f, "boolean"),
            ValueType::Json => write!(f, "json"),
        }
    }
}

/// Constraint operator
pub type ConstraintOperator = String;

/// Evaluation context passed per-evaluation
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EvaluationContext {
    pub user_id: Option<String>,
    pub session_id: Option<String>,
    pub app_name: Option<String>,
    pub app_version: Option<String>,
    pub remote_address: Option<String>,
    #[serde(default)]
    pub properties: HashMap<String, serde_json::Value>,
}

/// Strategy parameters
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StrategyParameters {
    pub rollout: Option<f64>,
    pub stickiness: Option<String>,
    pub group_id: Option<String>,
}

/// Constraint definition
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Constraint {
    pub context_name: String,
    pub operator: ConstraintOperator,
    pub value: Option<String>,
    pub values: Option<Vec<String>>,
    pub case_insensitive: Option<bool>,
    pub inverted: Option<bool>,
}

/// Variant definition
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Variant {
    pub name: String,
    pub weight: i32,
    pub value: Option<serde_json::Value>,
    #[serde(default)]
    pub enabled: bool,
}

/// Feature segment
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FeatureSegment {
    pub name: String,
    pub constraints: Vec<Constraint>,
}

/// Feature strategy
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FeatureStrategy {
    pub name: String,
    pub parameters: Option<StrategyParameters>,
    pub constraints: Option<Vec<Constraint>>,
    pub segments: Option<Vec<String>>,
    pub is_enabled: bool,
}

/// Feature flag definition
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FeatureFlag {
    pub id: String,
    pub name: String,
    pub is_enabled: bool,
    #[serde(default)]
    pub strategies: Vec<FeatureStrategy>,
    #[serde(default)]
    pub variants: Vec<Variant>,
    pub value_type: Option<String>,
    pub enabled_value: Option<serde_json::Value>,
    pub disabled_value: Option<serde_json::Value>,
    pub value_source: Option<String>,
}

/// Evaluation reason
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EvaluationReason {
    Disabled,
    Default,
    StrategyMatch,
    NotFound,
}

impl std::fmt::Display for EvaluationReason {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            EvaluationReason::Disabled => write!(f, "disabled"),
            EvaluationReason::Default => write!(f, "default"),
            EvaluationReason::StrategyMatch => write!(f, "strategy_match"),
            EvaluationReason::NotFound => write!(f, "not_found"),
        }
    }
}

/// Evaluation result
#[derive(Debug, Clone)]
pub struct EvaluationResult {
    pub id: String,
    pub flag_name: String,
    pub enabled: bool,
    pub reason: EvaluationReason,
    pub variant: Option<Variant>,
}

/// Evaluation detail with typed value
#[derive(Debug, Clone)]
pub struct EvaluationDetail<T> {
    pub value: T,
    pub reason: EvaluationReason,
    pub variant_name: Option<String>,
    pub flag_name: String,
}

/// Value source constants
pub mod value_source {
    pub const FLAG_DEFAULT_ENABLED: &str = "flag:defaultEnabled";
    pub const FLAG_DEFAULT_DISABLED: &str = "flag:defaultDisabled";
    pub const ENV_DEFAULT_ENABLED: &str = "env:defaultEnabled";
    pub const ENV_DEFAULT_DISABLED: &str = "env:defaultDisabled";
    pub const MISSING: &str = "missing";
}

/// Feature flags API response
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FeatureFlagsApiResponse {
    pub flags: Vec<FeatureFlag>,
    pub segments: Option<Vec<FeatureSegment>>,
    pub project_id: Option<String>,
}

/// Flag metric for analytics tracking
#[derive(Debug, Clone)]
pub struct FlagMetric {
    pub environment_id: String,
    pub flag_name: String,
    pub enabled: bool,
    pub variant_name: Option<String>,
    pub timestamp: chrono::DateTime<chrono::Utc>,
}
